### PR TITLE
support multiple shas

### PR DESCRIPTION
**Pull request by jmelis**
Part of https://issues.redhat.com/browse/APPSRE-2608

This is a major PR that introduces a major feature and modifies the internal mechanics of the server.

For consistent access, qontract-server clients can connect to it via a specific `<sha>` using this url: `/graphqlsha/<sha>`. If the data is reloaded the `sha` will also be renewed, and the server will start returning `409` indicating that there is a conflict and that it needs to fetch a new `sha`.

This has a negative impact on the reconciliation times for the integrations that query this server. Every time the data is reloaded, the integration is interrupted and has to start afresh. When the data is changed very frequently, the state may never be reconciled.

The goal of this PR is to add support for multiple shas. Each time the data is reloaded, the new bundle is exposed on its new `<sha>` (with the same well-known url `/graphqlsha/<sha>` thus maintaining backwards compatibility), but previous shas should continue to work. Note that `/graphql` will consistently point at the latest bundle, which is useful for casual access to the server, not for integrations.

**Cache Invalidation**

In order to guarantee memory efficiency, the shas will expire after a certain amount of time:

- When the data is loaded for the first time, the expiration time is set to 20 minutes in the future (can be overriden by the `BUNDLE_SHA_TTL` environment variable).
- Each time a sha is queried specifically the expiration is refreshed to the `BUNDLE_SHA_TTL` in the future again. This means that shas can be kept available forever by querying them before the `BUNDLE_SHA_TTL` has passed.
- The latest sha, which is the one pointed at by `/graphql` will never expire.
- Shas are only expired when `/reload` is queried.
- If `/reload` is called and there is no new data available, then no shas will be expired.

**Method changes**

- `GET /sha256`: (existed previously) Now it specifically points to the latest bundle. No backwards compatibility conflicts.
- `GET /git-commit`: (existed previously) Now it specifically points to the latest bundle. No backwards compatibility conflicts.
- `GET /git-commit/:sha`: (new method) It is possible to obtain now the git-commit for a specific sha.
- `GET /cache`: (new method) Prints the cache state.

**Metrics**

The metrics have been refactored to a separate file `metrics.ts`.

There are also new metrics introduced:

```
# HELP qontract_server_router_stack Number of layers in the router stack
# TYPE qontract_server_router_stack gauge

# HELP qontract_server_cache_bundle Number of shas cached by the application in the bundle object
# TYPE qontract_server_cache_bundle gauge

# HELP qontract_server_cache_bundle_cache Number of shas cached by the application in the bundleCache object
# TYPE qontract_server_cache_bundle_cache gauge
```

These new metrics are useful in order to ensure that there are no memory leaks. Note that `qontract_server_cache_bundle` and `qontract_server_cache_bundle_cache` should always coincide in number.

**Breaking changes**

There are **no breaking changes**. This PR preserves backwards compatibility. However, there are two minor changes that are worth mentioning.

- The server will no longer send `409` on any events. `409` indicates conflict, which is something that no longer applies as the shas will continue to be available. If the client queries a non-existing sha, or an expired one, it will receive `404`.

- ~~`GET /graphql` which was undocumented and just a corner case of the poorly defined middleware that existed previously will **no longer redirect** to `POST /graphql`. This was never the intention, so it has been removed. The tests however have been rewritten to properly send POST data, instead of performing GET queries as they were incorrectly doing before~~.

**Testing**

The functionality of the previously existing tests has remained untouched (except for renaming the `clusters` directory to `schemas` which is the proper name).

A new set of tests have been included in this PR which test the multiple shas functionality, and the bundle sha expiration and invalidation:

```
  multishas
    ✓ serves a basic graphql query (38ms)
    ✓ serves a basic graphql query using GET
Skipping reload, data already loaded
    ✓ reloads and works
reloaded
    ✓ loads new data and works
    ✓ access via sha1
    ✓ access via sha2
Removing expired bundle: 038f5149906c198069b08285bfba5a955531780793055a64a2bebc4920385279
reloaded
    ✓ removes expired bundle
    ✓ access via sha2
    ✓ access via sha3
```

**Things to look out for**

- Memory leaks. An important part of this PR resides in the logic that removes the expired shas. There is a new set of tests that verify that data is properly purged when the bundle shas are removed. However, this needs to be carefully reviewed.

- Every time that a new bundle is added it is mounted into the express server with this directive: `app.use(serverMiddleware);`. This is an `express` framework method call that pushes the middleware to the internal router stack. The challenge is that there is no documented / supported way to remove a middleware instance, which is precisely what we want to do when bundles expire. What we have done in this case is to access the private router stack (`app._router.stack`) and splice it in order to remove the relevant middleware. This is somewhat dangerous and may break if the `express` framework is upgraded (regardless if it's a major or minor update).

- The ApolloServer package has been upgraded. This is because the schema hot-reload internals have changed. This specifically affects the GraphQL inline fragments. However, this was a minor version upgrade, and it didn't affect the tests, so it is of minor importance.

EDIT: Reimplemented GET /graphql -> POST /graphql

### CODE REVIEW

**Review by maorfr — APPROVED** [review]
from just reading the description i understand that you thought of every scenario i could think of.
this is amazing work.

LGTM

**Comment by maorfr** [comment]
qontract-reconcile PR: https://github.com/app-sre/qontract-reconcile/pull/1135

File: test/multishas/multishas.test.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  it('access via sha3', async () => {
+    const sha: string = Object.keys(app.get('bundles'))[1];
+    const resp = await gql(srv, query2, sha);
+    resp.should.have.status(200);
+    resp.body.data.resources_v1[0].name.should.equal('sha3');
+    resp.body.data.resources_v1[0].resourceAField.should.equal('sha3');
+    resp.body.data.resources_v1[0].newfield.should.equal('sha3');
+  });
+
+});
```
Comment (rporres) [inline test/multishas/multishas.test.ts:138]: missing test on bad bundle reload?

**Review by rporres — CHANGES_REQUESTED** [review]
This is an impressive job. A few comments and suggestions here and there...

One thing I don't particularly like is breaking changes... Especially since we're doing `GET`s with `query` params, e.g. https://github.com/openshift/aws-account-operator/pull/493/files, and it makes sense that we're able to do a select with a query string via `GET` (on top of the `POST` support, I'd do both)

**Comment by jmelis** [comment]
> One thing I don't particularly like is breaking changes... Especially since we're doing GETs with query params, e.g. https://github.com/openshift/aws-account-operator/pull/493/files, and it makes sense that we're able to do a select with a query string via GET (on top of the POST support, I'd do both)

Reimplemented and updated PR description.

File: src/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+});
+
+const datafilesGuage = new promClient.Gauge({
+  name: 'qontract_server_datafiles',
+  help: 'Number of datafiles for a specific schema',
+  labelNames: ['schema'],
+});
+
+const routerStackGauge = new promClient.Gauge({
+  name: 'qontract_server_router_stack',
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/metrics.ts]: done

File: src/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  help: 'Number of layers in the router stack',
+});
+
+const bundleGauge = new promClient.Gauge({
+  name: 'qontract_server_cache_bundle',
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/metrics.ts]: done

File: src/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  help: 'Number of shas cached by the application in the bundle object',
+});
+
+const bundleCacheGauge = new promClient.Gauge({
+  name: 'qontract_server_cache_bundle_cache',
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/metrics.ts]: done

File: src/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      acc[d.$schema] = 0;
+    }
+    acc[d.$schema] += 1;
+    return acc;
+  };
+  const schemaCount: IAcct = bundle.datafiles.reduce(reducer, {});
+
+  // Set the Guage based on counted metrics
+  Object.keys(schemaCount).map(schemaName =>
+    datafilesGuage.set({ schema: schemaName }, schemaCount[schemaName]),
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/metrics.ts]: done

File: src/metrics.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  const reducer = (acc: IAcct, d: any) => {
+    if (!(d.$schema in acc)) {
+      acc[d.$schema] = 0;
+    }
+    acc[d.$schema] += 1;
+    return acc;
+  };
+  const schemaCount: IAcct = bundle.datafiles.reduce(reducer, {});
+
+  // Set the Guage based on counted metrics
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/metrics.ts]: done

File: src/server.ts [hunk 7]
```diff
@@ -1,162 +1,204 @@
...
 
 import * as db from './db';
+import * as metrics from './metrics';
 import { generateAppSchema, defaultResolver } from './schema';
 
-import promClient = require('prom-client');
-const promBundle = require('express-prom-bundle');
-
-import { GraphQLSchema } from 'graphql';
+// sha expiration time (in ms). Defaults to 1h.
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/server.ts]: done

File: src/server.ts [hunk 8]
```diff
@@ -1,162 +1,204 @@
...
-
-      req.app.set('bundle', bundle);
+    let bundle: db.Bundle;
 
-      const schema: GraphQLSchema = generateAppSchema(req.app as express.Express);
+    try {
+      // fetch the new bundle
+      bundle = await db.bundleFromEnvironment();
+    } catch (e) {
+      res.status(503).send('error parsing bundle, not replacing');
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/server.ts:143]: No, it just means that no new data will be loaded. The server will continue working as if `/reload` had not been called.

File: test/multishas/multishas.test.ts (hunk 1 above)
Comment (jmelis) [reply test/multishas/multishas.test.ts:159]: Good idea. Implemented.

File: src/server.ts [hunk 9]
```diff
@@ -1,162 +1,204 @@
...
-  labelNames: ['schema'],
-});
-
-const readFile = util.promisify(fs.readFile);
-
-export const appFromBundle = async (bundle: Promise<db.Bundle>) => {
+// registers a new ApolloServer into the app router and cache
+const registerApolloServer = (app: express.Express, bundleSha: string, server: any) => {
+  const serverMiddleware = server.getMiddleware({ path: `/graphqlsha/${bundleSha}` });
+  const expiration = Date.now() + BUNDLE_SHA_TTL;
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/server.ts:21]: I'm not sure I understand! can you elaborate?

File: src/server.ts [hunk 10]
```diff
@@ -1,162 +1,204 @@
...
+    }
+
+    const cacheInfo = cacheInfoObj as ICacheInfo;
+    if (cacheInfo.expiration < Date.now()) {
+      // removing sha
+      console.log(`Removing expired bundle: ${sha}`);
+      delete app.get('bundles')[sha];
+
+      // remove from router. NOTE: this is not officially supported and may break in future
+      // versions of express without warning.
```
[the remark this answers is not in the document]
Comment (jmelis) [reply src/server.ts:74]: Good idea. Just created this issue: https://github.com/expressjs/express/issues/4436

File: src/server.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rporres) [reply src/server.ts:21]: If this was a cache outside the process, you would want to keep the timestamp where it was created instead of the expiration, because if you change the value of the TTL you would have a lot of objects with an outdated expiration date. But in this case the discussion is academic, because the only way to change the TTL is restarting the server which will throw away all the objects 😄

File: src/metrics.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  help: 'Number of layers in the router stack',
+});
+
+const bundleGauge = new promClient.Gauge({
+  name: 'qontract_server_bundle_object_shas',
+  help: 'Number of shas cached by the application in the bundle object',
+});
+
+const bundleCacheGauge = new promClient.Gauge({
+  name: 'qontract_server_bundle_object_shas_cache',
```
Comment (rporres) [inline src/metrics.ts:48]: why not `qontract_server_bundle_cache_object_shas`?

**Comment by rporres** [comment]
And don't hate me too much, please convert the relevant bits of the description of this PR into proper documentation.

File: src/server.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmelis) [reply src/server.ts:21]: agreed!

File: src/metrics.ts (hunk 11 above)
Comment (jmelis) [reply src/metrics.ts]: fixed

**Comment by jmelis** [comment]
@rporres I have updated the README.md, it has been neglected so thanks for pointing it out.

**Review by rporres — APPROVED** [review]
Great stuff
